### PR TITLE
fix: use enveloped encoding for typed transactions

### PR DESCRIPTION
**Pull request by Rjected**
## Motivation
The encoding for `eth_sendRawTransaction` for typed transactions was not correct, because it was encoding a RLP header for typed transactions.

## Solution
This completely overhauls transaction encoding and decoding in alloy.

There is no longer `encode_signed` and `decode_signed` in `Transaction`. `Signed<T>` transactions cannot be encoded or decoded directly. Now, the main way to encode and decode transactions is through `Encodable2718` and `Decodable2718`, specifically through `TxEnvelope`.

The `TaggedLegacy` variant is removed.
`impl<T> std::ops::Deref for Signed<T>` is removed.

`impl From<Signed<TxEip4844WithSidecar>> for TxEnvelope` is added.
`impl From<Signed<TxEip4844>> for TxEnvelope` is added.

The 2718 trait methods now use `alloy_rlp::Result` instead of `Result<_, Eip2718Error>`

Raw RLP of a blob transaction with sidecar is added to `consensus/testdata`, and is used in a roundtrip test.

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Comment by Rjected** [comment]
fixing conflicts and adding blob tx with sidecar impl...

**Review by Evalir — COMMENTED** [review]
Awesome! Agreed we need to iron these out—but I believe enveloped encoding is already handled by `TxEnvelope` & `Decodable2718`/`Encodable2718`? (see `envelope.rs` & `alloy_eips::eip2718`)

**Comment by prestwich** [comment]
encode_signed and decode_signed probably should be documented as "dont call this, you want to call TxEnvelope methods instead"

File: crates/consensus/src/transaction/legacy.rs
Comment (prestwich) [inline crates/consensus/src/transaction/legacy.rs:96]: side note: i consider these all tech debt that is paid down by a refactor of rlp derives

Comment (Rjected) [reply crates/consensus/src/transaction/legacy.rs:96]: yes, rlp(flatten) etc will usher in a new era of derived and correct encodings

**Comment by gakonst** [comment]
@Rjected Plans here? Flagging that it seems like https://github.com/foundry-rs/foundry/pull/7242#issuecomment-1974253356 is blocked on this

**Comment by Rjected** [comment]
> @Rjected Plans here? Flagging that it seems like [foundry-rs/foundry#7242 (comment)](https://github.com/foundry-rs/foundry/pull/7242#issuecomment-1974253356) is blocked on this

I can get this rfr / done today to unblock

**Comment by Rjected** [comment]
Marked this as RFR, the code is ready to review, i plan on adding more tests today

File: crates/consensus/src/transaction/mod.rs
Comment (onbjerg) [inline crates/consensus/src/transaction/mod.rs:112]: why was this removed? it's useful for converting from mempool rlp -> typed tx

File: crates/consensus/src/transaction/mod.rs
Comment (onbjerg) [inline crates/consensus/src/transaction/mod.rs:106]: do we not need this?

File: crates/consensus/src/signed.rs
Comment (onbjerg) [inline crates/consensus/src/signed.rs:62]: i assume these were replaced by manual impls? if so, we should also note to implementers of `SignedTransaction` that they need to provide these by themselves, since it is needed to convert mempool tx -> typed tx

another q here: is it even possible to impl this in an external crate because of orphan rule? if not, we need to keep this

File: crates/consensus/src/transaction/mod.rs
Comment (onbjerg) [reply crates/consensus/src/transaction/mod.rs:112]: i see in the pr description that this is possible through TxEnvelope, could you give me a small example? we need to note this down for network crate authors, since it is likely a very important feature

File: crates/consensus/src/transaction/mod.rs
Comment (Rjected) [reply crates/consensus/src/transaction/mod.rs:106]: `encode_signed` and `decode_signed` are not specific enough for eth transactions, it's ambiguous whether or not you want the network or RPC RLP format for transactions. Encodable2718 provides this distinction

Comment (onbjerg) [reply crates/consensus/src/transaction/mod.rs:106]: ok, makes sense to me. need to document this, but can be a follow up, would create an issue

File: crates/consensus/src/signed.rs
Comment (onbjerg) [reply crates/consensus/src/signed.rs:62]: i see we can't blanket impl this anymore

**Review by onbjerg — APPROVED** [review]
lgtm

**Review by prestwich — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
rlp 👍

**Pull request merged**
